### ISSUE TITLE
Stabilizer state missing from sLSTM

### ISSUE THREAD

**Issue by Strawl** [muditbhargava66/PyxLSTM#15]
The new sLSTM doesn't have the stabilizer state m. This leads to exploding gradients very easily.

**Comment by strongerux** [comment]
Example: in slstm.py file
Chang this:
        for t in range(seq_length):
            x = input_seq[:, t, :]
            new_hidden_state = []
            for i, (lstm, f_gate, i_gate) in enumerate(zip(self.lstms, self.exp_forget_gates, self.exp_input_gates)):
                if hidden_state[i][0] is None:
                    h, c = lstm(x)
                else:
                    h, c = lstm(x, (hidden_state[i][0], hidden_state[i][1]))

                f = torch.exp(f_gate(h))
                i_exp = torch.exp(i_gate(h))
                c = f * c + i_exp * lstm.weight_hh.new_zeros(batch_size, self.hidden_size)
                new_hidden_state.append((h, c))

                if i < self.num_layers - 1:
                    x = self.dropout_layers[i](h)
                else:
                    x = h
            hidden_state = new_hidden_state
            output_seq.append(x)
to this:
        for t in range(seq_length):
            x = input_seq[:, t, :]
            new_hidden_state = []
            for i, (lstm, f_gate, i_gate, m_gate) in enumerate(zip(self.lstms, self.exp_forget_gates, self.exp_input_gates, self.exp_m_gates)):
                h_prev, c_prev, n_prev, m_prev = hidden_state[i]
                h, c = lstm(x, (h_prev, c_prev))
                f = torch.exp(f_gate(h))
                i_exp = torch.exp(i_gate(h))

                m = torch.max(torch.log(f) + m_prev, torch.log(i_exp))
                i_stabilized = torch.exp(torch.log(i_exp) - m)
                f_stabilized = torch.exp(torch.log(f) + m_prev - m)

                c = f_stabilized * c_prev + i_stabilized * lstm.weight_hh.new_zeros(batch_size, self.hidden_size)
                n = f_stabilized * n_prev + i_stabilized

                h = h * torch.sigmoid(lstm.weight_ih.new_ones(batch_size, self.hidden_size))
                new_hidden_state.append((h, c, n, m))

                if i < self.num_layers - 1:
                    x = self.dropout_layers[i](h)
                else:
                    x = h
            hidden_state = new_hidden_state
            output_seq.append(x)

        output_seq = torch.stack(output_seq, dim=1)
        return output_seq, hidden_state

**Comment by yanpeng0520** [comment]
Also the normalizer is not implemented?
